### PR TITLE
FEAT: Added a Ranking Chart

### PR DESCRIPTION
**Pull request by EncryptEx**
Added a Line chart that displays player points based on time.
Preview: 
Chart Preview

### CODE REVIEW

File: ranking/templates/ranking.html [hunk 1]
```diff
@@ -1,36 +1,111 @@
 {% extends 'base.html' %}
 {% load bootstrap3 %}
 {% load render_table from django_tables2 %}
+{% block head %}
+<!-- Required Libraries for Chart.js -->
+<script src="https://cdn.jsdelivr.net/npm/chart.js@4.3.2/dist/chart.umd.min.js"></script>
+<script src="https://code.jquery.com/jquery-3.6.4.min.js"
```
Comment (Casassarnau) [inline ranking/templates/ranking.html:7]: jQuery already added on thegame/base.html. Upgrade the version if you need it.

File: ranking/templates/ranking.html [hunk 2]
```diff
@@ -1,36 +1,111 @@
...
+                // Reset the current chart
+                chart.data.datasets = [];
+                chart.data.labels = [];
+
+                // Load new data into the chart
+                chart.options.title.text = title;
+                chart.options.title.display = true;
+
+                // Loop though every user (or dataset) and append it to chart
+                for (var key in datasets) {
```
Comment (Casassarnau) [inline ranking/templates/ranking.html:54]: It is more recomended to use let, not var.

File: readme.md [hunk 3]
```diff
@@ -30,7 +32,15 @@ Needs python 3.X and virtualenv
 - **name:** Challenge title
 - **description:** Challenge description (HTML format optional).
 - **order:** Challenge phase. User cannot see challenges if not completed all the challenges from previous phase (initial = 1).
-- **file:** Challenge file image, video, audio and HTML (django tags can be used).
+- **file:** Challenge file image,  Preparing metadata (setup.py) ... error
+  error: subprocess-exited-with-error
```
Comment (Casassarnau) [inline readme.md:36]: Is this supposed to be here? 🤔🤔

File: ranking/views.py [hunk 4]
```diff
@@ -13,3 +14,42 @@ class RankingView(LoginRequiredMixin, SingleTableView):
     def get_queryset(self):
         return ChallengeUser.objects.filter(success=True).values('user__username')\
             .annotate(count=Count('*'), time=Max('last_try')).order_by('-count', 'time')
+
+
+def get_chart(request):
```
Comment (Casassarnau) [inline ranking/views.py:19]: This endpoint is not secure. I would recommend to use a class that inherits the mixins of RankingView, then use the method get to put this code. (It works the same but with the mixins it has build in security).

File: ranking/views.py [hunk 5]
```diff
@@ -13,3 +14,42 @@ class RankingView(LoginRequiredMixin, SingleTableView):
...
+    # Get all Successful attempts
+    allEntries = list(ChallengeUser.objects.filter(success=True).values('user__username', 'last_try', 'total_attempts')
+                      .order_by('last_try'))
+
+    players = {}
+    Pcounter = {}
```
Comment (Casassarnau) [inline ranking/views.py:25]: Variables don't usually start in uppercase. In Phthon only classes start with uppercase.

File: ranking/views.py [hunk 6]
```diff
@@ -13,3 +14,42 @@ class RankingView(LoginRequiredMixin, SingleTableView):
...
+    top10 = ChallengeUser.objects.filter(success=True).values('user__username')\
+        .annotate(count=Count('*'), time=Max('last_try')).order_by('-count', 'time')[:10]
+    # Generate a whitelist of tthe top 10 usernames
+    top10names = [i['user__username'] for i in top10]
+
+    # Format the output required from ChartJS in order to work
+    final = []
+    for username in players:
+        if (username not in top10names):
+            continue
```
Comment (Casassarnau) [inline ranking/views.py:43]: Don't use continue, it makes code hard to read

**Review by Casassarnau — CHANGES_REQUESTED** [review]

File: readme.md (hunk 3 above)
Comment (EncryptEx) [reply readme.md]: Nope, thanksl

File: ranking/views.py (hunk 5 above)
Comment (EncryptEx) [reply ranking/views.py]: Fixed, thanks 4 the reminder

File: ranking/views.py (hunk 6 above)
Comment (EncryptEx) [reply ranking/views.py]: Thought guard clauses where more optimal

**Review by Casassarnau — APPROVED** [review]
LGTM

**Pull request merged**